src/app/config/resolve_addr.c

```diff
@@ -36,6 +61,469 @@ reset_last_resolved_addr_v4(void)
+ *     with it, the method_out is set to "CONFIGURED" and hostname_out is set
+ *     to NULL.
+ *
+ *     If Address is an internal address but NO custom authorities are used,
+ *     an error is returned.
+ *
+ *     If Address is a hostname, that is it can't be converted to an address,
+ *     it is resolved. On success, addr_out is set with the address,
+ *     method_out is set to "RESOLVED" and hostname_out is set to the resolved
+ *     hostname. On failure to resolve, an error is returned.
+ *
+ *     If no given Address, fallback to the local hostname (see section 2).
+ *
+ *  2. Look at the local hostname.
+ *
+ *     If the local hostname resolves to a non internal address, addr_out is
+ *     set with it, method_out is set to "GETHOSTNAME" and hostname_out is set
+ *     to the resolved hostname.
+ *
+ *     If a local hostname can NOT be found, an error is returned.
+ *
+ *     If the local hostname resolves to an internal address, an error is
+ *     returned.
+ *
+ *     If the local hostname can NOT be resolved, fallback to the network
+ *     interface (see section 3).
+ *
+ *  3. Look at the network interface.
+ *
+ *     Attempt to find the first public usable address from the list of
+ *     network interface returned by the OS.
+ *
+ *     On failure, an error is returned. This error indicates that all
+ *     attempts have failed and thus the address for the given family can not
+ *     be found.
+ *
+ *     On success, addr_out is set with it, method_out is set to "INTERFACE"
+ *     and hostname_out is set to NULL.
+ *
+ * @param options Global configuration options.
+ * @param family IP address family. Only AF_INET and AF_INET6 are supported.
+ * @param warn_severity Logging level.
+ * @param addr_out OUT: Set with the IP address found if any.
+ * @param method_out OUT: (optional) String denoting by which method the
+ *                   address was found. This is described in the
+ *                   control-spec.txt as actions for "STATUS_SERVER".
+ * @param hostname_out OUT: String containing the hostname if any was used.
+ *                     Only be set for "RESOLVED" and "GETHOSTNAME" methods.
+ *                     Else it is set to NULL.
+ *
+ * @return True if the address was found for the given family. False if not or
+ *         on errors.
+ */
+bool
+find_my_address(const or_options_t *options, int family, int warn_severity,
+                tor_addr_t *addr_out, const char **method_out,
+                char **hostname_out)
+{
+  int ret;
+  const char *method_used;
+  char *hostname_used = tor_strdup("");
+  tor_addr_t my_addr;
+
+  tor_assert(options);
+  tor_assert(addr_out);
+
+  /*
+   * Step 1: Discover address by attempting 3 different methods consecutively.
+   */
+
+  /* Attempt #1: Get address from configuration. */
+  ret = get_address_from_config(options, warn_severity, family, &method_used,
+                                &hostname_used, &my_addr);
+  if (ret == 0) {
```

Comment:
This whole if-then statement still seems very complex.  I wonder if we could refactor it into a loop over a function table?  We would need to make all the `get_address_from_*()` functions have the same API, but that might be a good thing.

Reply:
It is you are right. One of the main reason it took me a while to get this correctly. I actually had to reverse engineer the unit tests to learn a bunch of undocumented requirements that `resolve_my_address()` had...

Now one problem with a loop a set of functions is that the code right now behaves differently depending on the error code returned and which function it comes from.

Thus, we would also need a "what error code and what to do" table per function in this case. Or else we find a way to unify the error code. Maybe a function could return a "bail out" error, "keep going" error or "found" error code?

Reply:
That could make sense; or we could keep the current error codes but have another "what do I do about it value -- whatever you think would work best.

Reply:
Commit 7f1c9996163c6517